### ISSUE TITLE
Pods occasionally get killed (exit code 93) since upgrade to 3.3.6

### ISSUE THREAD

**Issue by derHeinzer** [argoproj/argo-workflows#8981]
## Checklist

 

* [x] Double-checked my configuration.
* [x] Tested using the latest version.
* [x] Used the Emissary executor.

## Summary

We recently upgraded from 3.2.3 to 3.3.6. Since the upgrade we are occassionally experiencing pods getting killed. We start ~180 pods all executing a python script, which trains and predicts some ML model with some variable arguments. We do this every day. Since the upgrade 1-3 pods get killed with following message:

/argo/staging/script: line 10:    93 Killed                  pyenv exec python script/execute.py "${args[@]}"

However the killed pods are not being deleted immediately. Instead it is being rerun after 75m (active deadline seconds 4500)  with respect to our configured retry policy. The second try has always succeeded for now.

In other occassions the return code might differ. In rare cases we get:
/argo/staging/script: line 10:    97 Killed                  pyenv exec python script/execute.py "${args[@]}"

First of all I am interested in what might cause this kills and what the return codes 93 and 97 mean?
Is it likely to be caused by the upgrade or some other component (k8s, Google Big query, scikit-learn, lightgbm, python version)? 

What version are you running?
-> 3.3.6

## Diagnostics

The pasted example configuration is not runnable without the python source code, which I am not able to provide. Nevertheless I hope every important information is provided with the partial configuration.

```yaml
    - name: train-predict-script
      inputs:
        parameters:
          - name: subcategory_partition_id
          - name: config_filename
          - name: cpu_request
            default: 2
          - name: memory_request
            default: 12Gi
          - name: expost
            default: "false"
          - name: job_timeout  # to be passed as activeDeadlineSeconds
            default: 4500  # 1 hour 15 minutes, must be lower than maxDuration of retry strategy
          - name: expost_mode_monthly
            default: "false"
      script:
        image: eu.gcr.io/{{workflow.parameters.projectId}}/psf:{{workflow.parameters.buildId}}
        command: [bash]
        source: |
          args=('-p={{workflow.parameters.projectId}}' \
          '-s={{inputs.parameters.subcategory_partition_id}}' \
          '-f={{inputs.parameters.config_filename}}' \
          '-b={{workflow.parameters.buildId}}')

          [[ {{inputs.parameters.expost}} == 'true' ]] && args+=( '-e' )
          [[ {{inputs.parameters.expost_mode_monthly}} == 'true' ]] && args+=('--expost-mode=monthly')
          echo "arguments: ${args[@]}"

          cd python/psf && ls -la && pyenv exec pip list && pyenv exec python script/execute.py "${args[@]}"
        imagePullPolicy: Always
        resources:
          requests:
            cpu: 2
            memory: 12Gi
      retryStrategy:
        limit: 2
        retryPolicy: "Always"
        backoff:
          duration: "5s"
          factor: 5
          maxDuration: "180m" # two activeDeadlineSeconds terminated runs are possible
      podSpecPatch: '{"activeDeadlineSeconds": {{inputs.parameters.job_timeout}}, "containers":[{"name":"main", "resources":{"requests":{"cpu": "{{inputs.parameters.cpu_request}}", "memory": "{{inputs.parameters.memory_request}}"}}}]}'

```

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.
[pod log.txt](https://github.com/argoproj/argo-workflows/files/8938794/pod.log.txt)
[pod-manifest-export-from-gui-with-small-comments.txt](https://github.com/argoproj/argo-workflows/files/8938795/pod-manifest-export-from-gui-with-small-comments.txt)
[controller log.zip](https://github.com/argoproj/argo-workflows/files/8938796/controller.log.zip)

**Comment by alexec** [comment]
I think your main container is being killed, but the wait container remains running.  I think exit code 93 is a red herring.

Can you try the `emissary` executor if you’re not using it, if you are can you try the `pns` executor?

After that, you please try running the `:latest` version of executor?

**Comment by derHeinzer** [comment]
We are using the emissary (default) executor in the :latest-version. Today we have been able to have a look in the container logs of one of the failed pods. It seems to be an OOM error, which is not argos fault at all. Nevertheless argo does not terminate the failing pod properly and the pod idles till the active deadline seconds thresh is reached. Besides a meaningfull error code and/or message would be highly appreciated.

**Comment by alexec** [comment]
Can you attach your controller logs and the full pod yaml?

**Comment by derHeinzer** [comment]
I attached 3 files:
1. the log of one failing pod. The pod log logs an OOM error (exit code 137), that does not show up in the argo gui. And the pod does not terminate but runs until it is restarted by configured active deadline seconds/retry policy.
2. full pod yaml
3. controller logs with some filters. The filters are:
resource.type="k8s_container"
resource.labels.cluster_name="flash"
resource.labels.namespace_name="argo"
labels.k8s-pod/app="workflow-controller"
-(textPayload=~"warning msg=\"workflow uses legacy/insecure pod patch") -> legacy rbac patching
-(textPayload=~"warning msg=\"Non-transient error") -> artifactory not found, but it works
-(textPayload=~"Get workflowtemplates 200")
-(textPayload=~"Enforcing history limit")
-(textPayload=~"info msg=\"Update leases 200")
-(textPayload=~"info msg=\"Get leases 200")
-(textPayload=~"info msg=\"Create events 201")

**Comment by juliev0** [comment]
This appears to be the same root cause as https://github.com/argoproj/argo-workflows/issues/9083, as mentioned by Alex above. In that bug, you can see my comment - I was able to show that the fix [here]([this](https://github.com/argoproj/argo-workflows/pull/8478/files) resolves this. Feel free to re-test with v3.4.0rc3 (using instructions on our [Release](https://github.com/argoproj/argo-workflows/releases) page) or otherwise using latest from master. Or we can just close this.
